src/flamenco/vm/fd_vm_interp_core.c

```diff
@@ -861,21 +867,70 @@
     reg[ dst ] = (ulong)( (uint)reg_dst % imm );
   FD_VM_INTERP_INSTR_END;
 
-  FD_VM_INTERP_BRANCH_BEGIN(0x95) /* FD_SBPF_OP_EXIT */
-      /* Agave JIT VM exit implementation analysis below.
-
-       Agave references:
-       https://github.com/solana-labs/rbpf/blob/v0.8.5/src/interpreter.rs#L503-L509
-       https://github.com/solana-labs/rbpf/blob/v0.8.5/src/jit.rs#L697-L702 */
-    if( FD_UNLIKELY( !frame_cnt ) ) goto sigexit; /* Exit program */
-    frame_cnt--;
-    reg[6]   = shadow[ frame_cnt ].r6;
-    reg[7]   = shadow[ frame_cnt ].r7;
-    reg[8]   = shadow[ frame_cnt ].r8;
-    reg[9]   = shadow[ frame_cnt ].r9;
-    reg[10]  = shadow[ frame_cnt ].r10;
-    pc       = shadow[ frame_cnt ].pc;
-  FD_VM_INTERP_BRANCH_END;
+  FD_VM_INTERP_BRANCH_BEGIN(0x95) { /* FD_SBPF_OP_CALL_IMM */
+    /* imm has already been validated to not overflow */
+    uint syscall_key = FD_VM_SBPF_STATIC_SYSCALLS_LIST[ imm ];
+    fd_sbpf_syscalls_t const * syscall = fd_sbpf_syscalls_query_const( syscalls, syscall_key, NULL );
+
+    /* this check is probably useless, as validation includes checking that the
+       syscall is active in this epoch.
+       However, it's safe to keep it here, because at the time of writing this
+       code we're not (re)validating all programs at every new epoch. */
+    if( FD_UNLIKELY( !syscall ) ) goto sigill;
+
+    /* Update the vm with the current vm execution state for the
+       syscall.  Note that BRANCH_BEGIN has pc at the syscall and
+       already updated ic and cu to reflect all instructions up to
+       and including the syscall instruction itself. */
+
+    vm->pc        = pc;
+    vm->ic        = ic;
+    vm->cu        = cu;
+    vm->frame_cnt = frame_cnt;
+
+    /* Do the syscall.  We use ret reduce the risk of the syscall
+       accidentally modifying other registers (note however since a
+       syscall has the vm handle it still do arbitrary modifications
+       to the vm state) and the risk of a pointer escape on reg from
+       inhibiting compiler optimizations (this risk is likely low in
+       as this is the only point in the whole interpreter core that
+       calls outside this translation unit). */
+
```

Comment:
All this logic is now duplicated from 0x85depr - is it worth consolidating these, so bug fixes in one make it into the other?

Reply:
added macro FD_VM_INTERP_SYSCALL_EXEC (too many var to make a function)